File: backend/internal/services/ai_usage_tracker.go
package services

import (
	"blog-backend/internal/database"
	"blog-backend/internal/models"
	"fmt"
	"log"
	"time"
)

// AIUsageTracker provides methods to track AI service usage
type AIUsageTracker struct{
	dailyCostLimit   float64
	monthlyCostLimit float64
}

// NewAIUsageTracker creates a new AI usage tracker instance
func NewAIUsageTracker() *AIUsageTracker {
	return &AIUsageTracker{
		dailyCostLimit:   1.0,  // $1 per day default limit
		monthlyCostLimit: 20.0, // $20 per month default limit
	}
}

// UsageMetrics contains metrics about an AI service call
type UsageMetrics struct {
	ServiceType   string
	Provider      string
	Model         string
	Operation     string
	InputTokens   int
	OutputTokens  int
	TotalTokens   int
	EstimatedCost float64
	Currency      string
	Language      string
	InputLength   int
	OutputLength  int
	ResponseTime  time.Duration
	Success       bool
	ErrorMessage  string
	ArticleID     *uint
	UserAgent     string
	IPAddress     string
}

// TrackUsage records AI service usage in the database with cost monitoring
func (tracker *AIUsageTracker) TrackUsage(metrics UsageMetrics) error {
	// Check cost limits before tracking
	if err := tracker.checkCostLimits(metrics.EstimatedCost); err != nil {
		log.Printf("⚠️ Cost limit warning: %v", err)
		// Don't block the operation, just warn
	}

	record := models.AIUsageRecord{
		ServiceType:   metrics.ServiceType,
		Provider:      metrics.Provider,
		Model:         metrics.Model,
		Operation:     metrics.Operation,
		InputTokens:   metrics.InputTokens,
		OutputTokens:  metrics.OutputTokens,
		TotalTokens:   metrics.TotalTokens,
		EstimatedCost: metrics.EstimatedCost,
		Currency:      metrics.Currency,
		Language:      metrics.Language,
		InputLength:   metrics.InputLength,
		OutputLength:  metrics.OutputLength,
		ResponseTime:  int(metrics.ResponseTime.Milliseconds()),
		Success:       metrics.Success,
		ErrorMessage:  metrics.ErrorMessage,
		ArticleID:     metrics.ArticleID,
		UserAgent:     metrics.UserAgent,
		IPAddress:     metrics.IPAddress,
	}

	return database.DB.Create(&record).Error
}

// GetUsageStats retrieves aggregated usage statistics
func (tracker *AIUsageTracker) GetUsageStats(serviceType, provider string, days int) ([]models.AIUsageStats, error) {
	var stats []models.AIUsageStats

	query := database.DB.Model(&models.AIUsageRecord{}).
		Select(`
			service_type,
			provider,
			COUNT(*) as total_requests,
			SUM(CASE WHEN success = 1 THEN 1 ELSE 0 END) as success_requests,
			SUM(total_tokens) as total_tokens,
			SUM(estimated_cost) as total_cost,
			currency,
			AVG(response_time) as avg_response_time
		`).
		Group("service_type, provider, currency")

	if days > 0 {
		query = query.Where("created_at >= ?", time.Now().AddDate(0, 0, -days))
	}

	if serviceType != "" {
		query = query.Where("service_type = ?", serviceType)
	}

	if provider != "" {
		query = query.Where("provider = ?", provider)
	}

	return stats, query.Scan(&stats).Error
}

// GetTotalCost calculates total cost for a time period
func (tracker *AIUsageTracker) GetTotalCost(days int) (float64, error) {
	var totalCost *float64

	query := database.DB.Model(&models.AIUsageRecord{}).
		Select("SUM(estimated_cost)")

	if days > 0 {
		query = query.Where("created_at >= ?", time.Now().AddDate(0, 0, -days))
	}

	err := query.Scan(&totalCost).Error
	if err != nil {
		return 0, err
	}

	// If no records exist, SUM returns NULL
	if totalCost == nil {
		return 0, nil
	}

	return *totalCost, nil
}

// GetRecentUsage retrieves recent usage records
func (tracker *AIUsageTracker) GetRecentUsage(limit int) ([]models.AIUsageRecord, error) {
	var records []models.AIUsageRecord

	return records, database.DB.
		Order("created_at DESC").
		Limit(limit).
		Find(&records).Error
}

// GetUsageByArticle retrieves usage records for a specific article
func (tracker *AIUsageTracker) GetUsageByArticle(articleID uint) ([]models.AIUsageRecord, error) {
	var records []models.AIUsageRecord

	return records, database.DB.
		Where("article_id = ?", articleID).
		Order("created_at DESC").
		Find(&records).Error
}

// GetDailyUsage retrieves daily usage statistics for the last N days
func (tracker *AIUsageTracker) GetDailyUsage(days int) (map[string]models.AIUsageStats, error) {
	var results []struct {
		Date            string  `json:"date"`
		TotalRequests   int64   `json:"total_requests"`
		SuccessRequests int64   `json:"success_requests"`
		TotalTokens     int64   `json:"total_tokens"`
		TotalCost       float64 `json:"total_cost"`
		AvgResponseTime float64 `json:"avg_response_time"`
	}

	err := database.DB.Model(&models.AIUsageRecord{}).
		Select(`
			DATE(created_at) as date,
			COUNT(*) as total_requests,
			SUM(CASE WHEN success = 1 THEN 1 ELSE 0 END) as success_requests,
			SUM(total_tokens) as total_tokens,
			SUM(estimated_cost) as total_cost,
			AVG(response_time) as avg_response_time
		`).
		Where("created_at >= ?", time.Now().AddDate(0, 0, -days)).
		Group("DATE(created_at)").
		Order("date DESC").
		Scan(&results).Error

	if err != nil {
		return nil, err
	}

	// Convert to map for easier access
	dailyStats := make(map[string]models.AIUsageStats)
	for _, result := range results {
		dailyStats[result.Date] = models.AIUsageStats{
			TotalRequests:   result.TotalRequests,
			SuccessRequests: result.SuccessRequests,
			TotalTokens:     result.TotalTokens,
			TotalCost:       result.TotalCost,
			AvgResponseTime: result.AvgResponseTime,
		}
	}

	return dailyStats, nil
}

// CleanupOldRecords removes records older than specified days (for data retention)
func (tracker *AIUsageTracker) CleanupOldRecords(days int) (int64, error) {
	cutoffDate := time.Now().AddDate(0, 0, -days)

	result := database.DB.Where("created_at < ?", cutoffDate).Delete(&models.AIUsageRecord{})
	return result.RowsAffected, result.Error
}

// checkCostLimits checks if adding this cost would exceed daily or monthly limits
func (tracker *AIUsageTracker) checkCostLimits(newCost float64) error {
	// Check daily cost
	dailyCost, err := tracker.GetTotalCost(1)
	if err != nil {
		return fmt.Errorf("failed to get daily cost: %v", err)
	}
	
	if dailyCost+newCost > tracker.dailyCostLimit {
		return fmt.Errorf("daily cost limit exceeded: current=$%.6f, limit=$%.6f, new request would add=$%.6f", 
			dailyCost, tracker.dailyCostLimit, newCost)
	}
	
	// Check monthly cost
	monthlyCost, err := tracker.GetTotalCost(30)
	if err != nil {
		return fmt.Errorf("failed to get monthly cost: %v", err)
	}
	
	if monthlyCost+newCost > tracker.monthlyCostLimit {
		return fmt.Errorf("monthly cost limit exceeded: current=$%.6f, limit=$%.6f, new request would add=$%.6f", 
			monthlyCost, tracker.monthlyCostLimit, newCost)
	}
	
	return nil
}

// GetCostSummary returns current cost summary and limits
func (tracker *AIUsageTracker) GetCostSummary() (map[string]interface{}, error) {
	dailyCost, err := tracker.GetTotalCost(1)
	if err != nil {
		return nil, err
	}
	
	monthlyCost, err := tracker.GetTotalCost(30)
	if err != nil {
		return nil, err
	}
	
	return map[string]interface{}{
		"daily": map[string]interface{}{
			"cost":       dailyCost,
			"limit":      tracker.dailyCostLimit,
			"percentage": (dailyCost / tracker.dailyCostLimit) * 100,
			"remaining":  tracker.dailyCostLimit - dailyCost,
		},
		"monthly": map[string]interface{}{
			"cost":       monthlyCost,
			"limit":      tracker.monthlyCostLimit,
			"percentage": (monthlyCost / tracker.monthlyCostLimit) * 100,
			"remaining":  tracker.monthlyCostLimit - monthlyCost,
		},
	}, nil
}

// SetCostLimits updates the cost limits
func (tracker *AIUsageTracker) SetCostLimits(dailyLimit, monthlyLimit float64) {
	tracker.dailyCostLimit = dailyLimit
	tracker.monthlyCostLimit = monthlyLimit
	log.Printf("💰 Updated cost limits: daily=$%.2f, monthly=$%.2f", dailyLimit, monthlyLimit)
}

File: frontend/src/services/ai-usage-tracker.ts
import { apiClient } from '../lib/api'

export interface AIUsageMetrics {
  serviceType: 'summary' | 'translation' | 'seo'
  provider: string
  model?: string
  operation: string
  inputTokens?: number
  outputTokens?: number
  totalTokens?: number
  estimatedCost?: number
  currency?: string
  language?: string
  inputLength?: number
  outputLength?: number
  responseTime?: number // milliseconds
  success: boolean
  errorMessage?: string
  articleId?: number
}

export interface AIUsageStats {
  serviceType: string
  provider: string
  totalRequests: number
  successRequests: number
  totalTokens: number
  totalCost: number
  currency: string
  avgResponseTime: number
}

export interface DailyUsageStats {
  date: string
  totalRequests: number
  successRequests: number
  totalTokens: number
  totalCost: number
  avgResponseTime: number
}

export interface CostLimits {
  dailyLimit: number
  monthlyLimit: number
}

export interface CostSummary {
  dailyLimit: number
  monthlyLimit: number
  summary: {
    daily: {
      cost: number
      limit: number
      percentage: number
      remaining: number
    }
    monthly: {
      cost: number
      limit: number
      percentage: number
      remaining: number
    }
  }
}

export class AIUsageTracker {
  private static instance: AIUsageTracker
  private trackingEnabled: boolean = true

  private constructor() {}

  static getInstance(): AIUsageTracker {
    if (!AIUsageTracker.instance) {
      AIUsageTracker.instance = new AIUsageTracker()
    }
    return AIUsageTracker.instance
  }

  setTrackingEnabled(enabled: boolean): void {
    this.trackingEnabled = enabled
  }

  async trackUsage(metrics: AIUsageMetrics): Promise<void> {
    if (!this.trackingEnabled) {
      return
    }

    try {
      // Convert camelCase to snake_case for backend API
      const payload = {
        service_type: metrics.serviceType,
        provider: metrics.provider,
        model: metrics.model,
        operation: metrics.operation,
        input_tokens: metrics.inputTokens || 0,
        output_tokens: metrics.outputTokens || 0,
        total_tokens: metrics.totalTokens || 0,
        estimated_cost: metrics.estimatedCost || 0,
        currency: metrics.currency || 'USD',
        language: metrics.language,
        input_length: metrics.inputLength || 0,
        output_length: metrics.outputLength || 0,
        response_time: metrics.responseTime || 0,
        success: metrics.success,
        error_message: metrics.errorMessage || '',
        article_id: metrics.articleId
      }

      await apiClient.request('/ai-usage/track', {
        method: 'POST',
        body: JSON.stringify(payload)
      })
    } catch (error) {
      console.error('Failed to track AI usage:', error)
      // Don't throw error to avoid breaking the main flow
    }
  }

  async getUsageStats(
    serviceType?: string, 
    provider?: string, 
    days: number = 30
  ): Promise<AIUsageStats[]> {
    try {
      const params = new URLSearchParams()
      if (serviceType) params.append('service_type', serviceType)
      if (provider) params.append('provider', provider)
      params.append('days', days.toString())

      const response = await apiClient.request<{ 
        stats: Array<{
          service_type: string
          provider: string
          total_requests: number
          success_requests: number
          total_tokens: number
          total_cost: number
          currency: string
          avg_response_time: number
        }>
      }>(`/ai-usage/stats?${params.toString()}`)
      
      // Convert snake_case to camelCase
      const convertedStats: AIUsageStats[] = response.stats.map(stat => ({
        serviceType: stat.service_type,
        provider: stat.provider,
        totalRequests: stat.total_requests,
        successRequests: stat.success_requests,
        totalTokens: stat.total_tokens,
        totalCost: stat.total_cost,
        currency: stat.currency,
        avgResponseTime: stat.avg_response_time
      }))
      
      return convertedStats
    } catch (error) {
      console.error('Failed to get usage stats:', error)
      throw error
    }
  }

  async getTotalCost(days: number = 30): Promise<{ totalCost: number; currency: string; periodDays: number }> {
    try {
      const response = await apiClient.request<{ total_cost: number; currency: string; period_days: number }>(`/ai-usage/cost?days=${days}`)
      return {
        totalCost: response.total_cost,
        currency: response.currency,
        periodDays: response.period_days
      }
    } catch (error) {
      console.error('Failed to get total cost:', error)
      throw error
    }
  }

  async getRecentUsage(limit: number = 50): Promise<any[]> {
    try {
      const response = await apiClient.request<{ records: any[] }>(`/ai-usage/recent?limit=${limit}`)
      return response.records
    } catch (error) {
      console.error('Failed to get recent usage:', error)
      throw error
    }
  }

  async getDailyUsage(days: number = 30): Promise<Record<string, DailyUsageStats>> {
    try {
      const response = await apiClient.request<{ daily_stats: Record<string, DailyUsageStats> }>(`/ai-usage/daily?days=${days}`)
      return response.daily_stats
    } catch (error) {
      console.error('Failed to get daily usage:', error)
      throw error
    }
  }

  async getUsageByArticle(articleId: number): Promise<any[]> {
    try {
      const response = await apiClient.request<{ records: any[] }>(`/ai-usage/article/${articleId}`)
      return response.records
    } catch (error) {
      console.error('Failed to get article usage:', error)
      throw error
    }
  }

  async getCostLimits(): Promise<CostSummary> {
    try {
      const response = await apiClient.request<{
        daily_limit: number
        monthly_limit: number
        summary: {
          daily: {
            cost: number
            limit: number
            percentage: number
            remaining: number
          }
          monthly: {
            cost: number
            limit: number
            percentage: number
            remaining: number
          }
        }
      }>('/ai-usage/cost-limits')
      
      return {
        dailyLimit: response.daily_limit,
        monthlyLimit: response.monthly_limit,
        summary: response.summary
      }
    } catch (error) {
      console.error('Failed to get cost limits:', error)
      throw error
    }
  }

  async setCostLimits(dailyLimit: number, monthlyLimit: number): Promise<void> {
    try {
      await apiClient.request('/ai-usage/cost-limits', {
        method: 'PUT',
        body: JSON.stringify({
          daily_limit: dailyLimit,
          monthly_limit: monthlyLimit
        })
      })
    } catch (error) {
      console.error('Failed to set cost limits:', error)
      throw error
    }
  }

  // Helper method to wrap AI operations with automatic tracking
  async trackOperation<T>(
    operation: () => Promise<T>,
    baseMetrics: Omit<AIUsageMetrics, 'success' | 'responseTime'>
  ): Promise<T> {
    const startTime = Date.now()
    let success = false
    let errorMessage: string | undefined
    let result: T

    try {
      result = await operation()
      success = true
      return result
    } catch (error) {
      errorMessage = error instanceof Error ? error.message : 'Unknown error'
      throw error
    } finally {
      const responseTime = Date.now() - startTime
      
      await this.trackUsage({
        ...baseMetrics,
        success,
        responseTime,
        errorMessage
      })
    }
  }
}

// Export singleton instance
export const aiUsageTracker = AIUsageTracker.getInstance()

// Helper functions for different AI operations
export async function trackSummaryGeneration<T>(
  operation: () => Promise<T>,
  provider: string,
  model: string,
  language: string,
  inputLength: number,
  articleId?: number
): Promise<T> {
  return aiUsageTracker.trackOperation(operation, {
    serviceType: 'summary',
    provider,
    model,
    operation: 'generate_summary',
    language,
    inputLength,
    articleId
  })
}

export async function trackTranslation<T>(
  operation: () => Promise<T>,
  provider: string,
  fromLanguage: string,
  toLanguage: string,
  inputLength: number,
  articleId?: number
): Promise<T> {
  return aiUsageTracker.trackOperation(operation, {
    serviceType: 'translation',
    provider,
    operation: 'translate_text',
    language: `${fromLanguage}->${toLanguage}`,
    inputLength,
    articleId
  })
}

export async function trackSEOGeneration<T>(
  operation: () => Promise<T>,
  provider: string,
  operationType: 'keywords' | 'title' | 'description',
  language: string,
  inputLength: number,
  articleId?: number
): Promise<T> {
  return aiUsageTracker.trackOperation(operation, {
    serviceType: 'seo',
    provider,
    operation: `generate_${operationType}`,
    language,
    inputLength,
    articleId
  })
}